File: python/42_TrappingRainWater.py
from unittest import TestCase
# https://leetcode.com/problems/trapping-rain-water/


class TrappingRainWater(TestCase):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        if not height: return 0
        sum = 0
        k = 0
        for i in range(1, len(height)):
            if height[i] > height[k]:
                k = i

        mx = height[0]
        for i in range(1, k):
            if height[i] > mx:
                mx = height[i]
            else:
                sum += mx - height[i]

        mx = height[-1]
        for i in range(len(height)-2, k, -1):
            if height[i] > mx:
                mx = height[i]
            else:
                sum += mx - height[i]

        return sum

    def test1(self):
        self.assertEqual(6, self.trap([0,1,0,2,1,0,1,3,2,1,2,1]))

File: kotlin/src/main/kotlin/42.TrappingRainWater.kt
import org.junit.Assert.assertEquals
import org.junit.Test

//https://leetcode.com/problems/trapping-rain-water/

class TrappingRainWater {
  fun trap(height: IntArray): Int {
    if(height.isEmpty()) return 0

    var sum = 0
    val k = height.withIndex().maxBy{it.value}!!.index

    var max = height.first()
    for(i in 1 until k)
      if(height[i] > max)
        max = height[i]
      else
        sum += max-height[i]

    max = height.last()
    for(i in height.size-1 downTo k+1)
      if(height[i] > max)
        max = height[i]
      else
        sum += max - height[i]

    return sum
  }

  @Test
  fun test1(){
    assertEquals(6, trap(intArrayOf(0,1,0,2,1,0,1,3,2,1,2,1)))
  }

  @Test
  fun test2(){
    assertEquals(9, trap(intArrayOf(4,2,0,3,2,5)))
  }
}